File: algorithms/medium/the_number_of_weak_characters_in_the_game.ex
defmodule Solution do
  @spec number_of_weak_characters(properties :: [[integer]]) :: integer
  def number_of_weak_characters(properties) do
    properties
    |> Enum.sort(fn [a0, a1], [b0, b1] ->
      if a0 == b0, do: b1 <= a1, else: a0 <= b0
    end)
    |> Enum.reverse()
    |> Enum.reduce({0, 0}, fn [_, d], {max, ans} ->
      if d < max, do: {max, ans + 1}, else: {d, ans}
    end)
    |> elem(1)
  end
end

File: algorithms/medium/the_number_of_weak_characters_in_the_game.rs
impl Solution {
    pub fn number_of_weak_characters(mut properties: Vec<Vec<i32>>) -> i32 {
        properties.sort_by(|a, b| {
            if a[0] == b[0] {
                b[1].cmp(&a[1])
            } else {
                a[0].cmp(&b[0])
            }
        });
        let mut max_defense = 0;
        let mut ans = 0;
        for p in properties.iter().rev() {
            if p[1] < max_defense {
                ans += 1;
            } else {
                max_defense = p[1];
            }
        }
        ans
    }
}
